### PR TITLE
Homework 28 and 29.

### PR DESCRIPTION
**Pull request by ESergiusz**

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -8,38 +8,87 @@ services:
...
 volumes:
   nginx_log:
     driver: local
     driver_opts:
       o: bind
       type: none
-      device: /var/log/nginx_logs
\ No newline at end of file
+      device: /var/log/nginx_logs
+  certbot_log:
```
Comment (ggramal) [inline docker-compose.yaml:77]: Hmm what kind of logs are written to this location :thinking: ?

File: ansible/roles/service_authz/tasks/main.yml [hunk 2]
```diff
@@ -1,6 +1,63 @@
...
+      ansible.builtin.file:
+        path: /etc/letsencrypt
+        state: directory
+        mode: '0755'
+    - name: Create a directory for SSL certificate (certbot)
+      ansible.builtin.file:
+        path: /etc/letsencrypt/live/esa.authz.smodata.net
+        state: directory
+        mode: '0755'
+    - name: Copy file certbot.sh
```
Comment (ggramal) [inline ansible/roles/service_authz/tasks/main.yml:25]: So we are using a separate container for letsencrypt. Why not add certbot.sh to our container image in Dockerfile,certbot

File: docker-compose.yaml [hunk 3]
```diff
@@ -8,38 +8,87 @@ services:
...
+    depends_on:
+      - nginx
+    networks:
+      app-network:
+        ipv4_address: 172.22.0.7
+  db:
+    image: postgres:latest
+    restart: always
+    env_file:
+      - /etc/PGDATA/db_login.env
```
Comment (ggramal) [inline docker-compose.yaml:55]: Can we use default  `.env` file here. So the approach is
* create 2 files `.env.example` `.env`
* store in `.env.example` env vars that need to be set and commit it
* add .env to .gitignore and do not commit it
This way 
* when devs are testing locally they can
   * copy `.env.example` -> `.env` and adjust env vars base on their needs
   * run docker-compose locally
* we can use ansible to encrypt   .env file with credentials related to our runtime environments like dev prod, pre and deploy it to them

File: docker-compose.yaml [hunk 4]
```diff
@@ -8,38 +8,87 @@ services:
...
+    driver_opts:
+      o: bind
+      type: none
+      device: /etc/letsencrypt
+  PGDATA:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /etc/PGDATA
```
Comment (ggramal) [inline docker-compose.yaml:94]: As a developer i want to run the application and all its dependencies  like dbs locally. docker-compose is the optimal tool for that. But for me running this compose i need to add folder/copy files etc.So what i propose is to use default docker volumes here without specifying `device` attribute

File: docker-compose.yaml (hunk 1 above)
Comment (ESergiusz) [reply docker-compose.yaml]: To see the logs of the certbot operation, if a new certificate is issued.

File: docker-compose.yaml (hunk 4 above)
Comment (ESergiusz) [reply docker-compose.yaml]: All comments are taken into account. I had to tinker.

File: conf/certbot/privkey.pem [hunk 5]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (ESergiusz) [reply conf/certbot/privkey.pem]: Fixed

**Comment by ggramal** [comment]
How should i run docker-compose locally?

File: docker-compose.yaml [hunk 6]
```diff
@@ -8,38 +8,65 @@ services:
...
+      - /var/authz/log/certbot_log:/var/log/letsencrypt
+    networks:
+      app-network:
+        ipv4_address: 172.22.0.7
+  db:
+    image: postgres:latest
+    container_name: Postgres
+    restart: always
+    env_file:
+      - /tmp/db_login.env
```
Comment (ggramal) [inline docker-compose.yaml:57]: how should i run this `docker-compose` locally

Comment (ggramal) [reply docker-compose.yaml:57]: + 
```
TASK [service_authz : Build container (docker-compose)] *******************************************************************************************
fatal: [3.123.23.80]: FAILED! => {"changed": false, "msg": "Configuration error - Couldn't find env file: /tmp/db_login.env"}
```

Comment (ESergiusz) [reply docker-compose.yaml:59]: Fixed. 
If you want to run this docker-compose locally, you need to run decrypt_secrets.sh before launching docker-compose
🤞

**Pull request merged**